crates/stackable-operator/src/utils/cluster_domain.rs

```diff
@@ -0,0 +1,183 @@
+    AmbiguousDomainEntries,
+}
+
+/// Tries to retrieve the Kubernetes cluster domain.
+///
+/// 1. Return `KUBERNETES_CLUSTER_DOMAIN` if set, otherwise
+/// 2. Return the cluster domain parsed from the `/etc/resolv.conf` file if `KUBERNETES_SERVICE_HOST`
+///    is set, otherwise fall back to `cluster.local`.
+///
+/// This variable is initialized in [`crate::client::initialize_operator`], which is called in the
+/// main function. It can be used as suggested below.
+///
+/// ## Usage
+///
+/// ```no_run
+/// use stackable_operator::utils::KUBERNETES_CLUSTER_DOMAIN;
+///
+/// let kubernetes_cluster_domain = KUBERNETES_CLUSTER_DOMAIN.get()
+///     .expect("KUBERNETES_CLUSTER_DOMAIN must first be set by calling initialize_operator");
+///
+/// tracing::info!(%kubernetes_cluster_domain, "Found cluster domain");
+/// ```
+///
+/// ## See
+///
+/// - <https://github.com/stackabletech/issues/issues/436>
+/// - <https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/>
+pub static KUBERNETES_CLUSTER_DOMAIN: OnceLock<DomainName> = OnceLock::new();
+
+#[instrument]
+pub(crate) fn retrieve_cluster_domain() -> Result<DomainName, Error> {
+    // 1. Read KUBERNETES_CLUSTER_DOMAIN env var
+    tracing::debug!("Trying to determine the Kubernetes cluster domain...");
+
+    match env::var(KUBERNETES_CLUSTER_DOMAIN_ENV) {
+        Ok(cluster_domain) if !cluster_domain.is_empty() => {
+            let cluster_domain = DomainName::from_str(&cluster_domain)
+                .context(ParseDomainNameSnafu { cluster_domain })?;
+            tracing::info!(
+                %cluster_domain,
+                "Using Kubernetes cluster domain from {KUBERNETES_CLUSTER_DOMAIN_ENV} environment variable"
+            );
+            return Ok(cluster_domain);
+        }
+        _ => {}
+    };
+
+    // 2. If no env var is set, check if we run in a clustered (Kubernetes/Openshift) environment
+    //    by checking if KUBERNETES_SERVICE_HOST is set: If not default to 'cluster.local'.
+    tracing::debug!(
+        "Trying to determine the operator runtime environment as environment variable \
+            \"{KUBERNETES_CLUSTER_DOMAIN_ENV}\" is not set"
+    );
+
+    match env::var(KUBERNETES_SERVICE_HOST_ENV) {
+        Ok(_) => {
+            let cluster_domain = retrieve_cluster_domain_from_resolv_conf(RESOLVE_CONF_FILE_PATH)?;
+            let cluster_domain = DomainName::from_str(&cluster_domain)
+                .context(ParseDomainNameSnafu { cluster_domain })?;
+
+            tracing::info!(
+                %cluster_domain,
+                "Using Kubernetes cluster domain from {RESOLVE_CONF_FILE_PATH} file"
+            );
+
+            Ok(cluster_domain)
+        }
+        Err(_) => {
+            let cluster_domain = DomainName::from_str(KUBERNETES_CLUSTER_DOMAIN_DEFAULT).context(
+                ParseDomainNameSnafu {
+                    cluster_domain: KUBERNETES_CLUSTER_DOMAIN_DEFAULT,
+                },
+            )?;
```

Comment:
As mentioned in comments on the error variants, we should emit a warning and fall back to the default.

```suggestion
            let cluster_domain = DomainName::from_str(KUBERNETES_CLUSTER_DOMAIN_DEFAULT).expect("KUBERNETES_CLUSTER_DOMAIN_DEFAULT is a valid domain");
```

Reply:
Done in c6865684dababfdb643dfbf2a5d062aa38514822